man/rvc2.1
.Dd Dec 28, 2020
.Dt RVC2 1
.Sh NAME
.Nm rvc2
.Nd radare version control
.Sh SYNOPSIS
.Nm rvc2
.Op Fl ebBsSvxkKh
.Op [expr] ...
.Sh DESCRIPTION
This command is part of the radare project.
.Pp
Implements a simple version control system that aims to work as frontend for git, as well as its own structure for versioning project changes.
.Pp
This command is pretty new and experimental.
.Sh OPTIONS
.It Fl h
Show usage help message
.It Fl v
Display version
.El
.Sh USAGE
Available variable types are:
.Pp
rvc2 init
echo hello world > README.md
rvc2 add README.md
rvc2 branch
rvc2 commit
.Pp
.Sh SEE ALSO
.Pp
.Xr radare2(1)
.Sh AUTHORS
.Pp
Written by pancake <pancake@nopcode.org>.


ChangeLog
-=== RELEASE 0.8 ===-
changeset:   1625:0aaad74ad9fc
bookmark:    master
tag:         tip
user:        pancake
date:        Fri Jul 15 18:13:00 2011 +0200
files:       configure configure.acr libr/core/core.c libr/io/undo.c r2-bindings/python/r_cons.i.experimental
description:
* Bump release 0.8 - Codename thecakeisapie
* Fix uninitialized memory issue in r_block_resize()
  - Thanks vext01 for reporting


changeset:   1624:1315774824fb
parent:      1622:fc2643c36e30
parent:      1623:95f09764122e
user:        pancake
date:        Fri Jul 15 18:07:56 2011 +0200
description:
* Merge heads


changeset:   1623:95f09764122e
parent:      1620:236b81674ab5
user:        earada
date:        Fri Jul 15 14:17:36 2011 +0200
files:       libr/bin/p/bin_elf.c
description:
* Fix virtual address in elf without setcions


changeset:   1622:fc2643c36e30
user:        pancake
date:        Fri Jul 15 12:57:30 2011 +0200
files:       mk/gitpush.sh
description:
* Fix some typos in gitpush script


changeset:   1621:b6a6d3012998
user:        pancake
date:        Fri Jul 15 12:48:46 2011 +0200
files:       Makefile mk/gitpush.sh
description:
* Added gitpush utility for github mirroring


changeset:   1620:236b81674ab5
user:        pancake@pair
date:        Fri Jul 15 02:21:40 2011 +0200
files:       configure configure.acr
description:
* Bump acr to 0.8.6 with shellscript endian check
  - New flags --with-{little|big}-endian


changeset:   1619:d49dd6f2e789
user:        pancake@pair
date:        Fri Jul 15 01:50:42 2011 +0200
files:       binr/Makefile doc/maemo libr/Makefile maemo/Makefile maemo/README maemo/deb_hand.mak maemo/description.txt
description:
* Added support for new platform Maemo6-Harmattan
  - Linux/arm with debugger support
  - Debian package generation tools (no need for dpkg)
  - Compilation instructions at doc/maemo
* Fix install in binr and libr


changeset:   1618:dd1c50f121bf
user:        pancake@pair
date:        Thu Jul 14 22:14:11 2011 +0200
files:       configure configure.acr
description:
* Handle -lssl and pkg-config openssl
  - Should fix NetBSD build


changeset:   1617:4c6e8c50e7cf
user:        earada
date:        Thu Jul 14 16:17:39 2011 +0200
files:       libr/debug/p/debug_native.c libr/util/sys.c
description:
* Fix FreeBSD build


changeset:   1616:fd36edc76e42
user:        pancake
date:        Wed Jul 13 19:37:28 2011 +0200
files:       test/t/asm
description:
* Fix asm test


changeset:   1615:1ad69f14f1db
user:        pancake
date:        Wed Jul 13 19:28:53 2011 +0200
files:       test/tests.sh
description:
* Fix test suite


changeset:   1614:6c552be332cb
user:        pancake
date:        Wed Jul 13 19:22:38 2011 +0200
files:       binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile config-user.mk.acr configure configure.acr libr/debug/t/Makefile libr/hash/t/Makefile libr/include/r_userconf.h.acr libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/t/Makefile libr/socket/Makefile libr/socket/t/Makefile
description:
* Use OpenSSL (cflags/ldflags) from pkg-config


changeset:   1613:8ecc34f29bc4
user:        pancake
date:        Wed Jul 13 18:55:16 2011 +0200
files:       doc/windows libr/asm/arch/msil/demsil.c libr/asm/p/asm_dalvik.c libr/bin/format/mach0/dyldcache.c libr/config.mk.tail libr/core/vmenus.c libr/include/list.h libr/include/r_types_base.h libr/io/p/io_w32dbg.c libr/socket/t/test.c libr/util/cache.c
description:
* Fix build and lot of warnings for w64


changeset:   1612:1552f58ae98c
user:        earada
date:        Wed Jul 13 17:41:26 2011 +0200
files:       binr/radare2/radare2.c libr/core/cmd.c libr/core/config.c libr/core/file.c libr/include/r_core.h
description:
* Use native api to calculate md5/sha1 hashes
- Fix bug 29


changeset:   1611:3dd0b00836fe
user:        pancake
date:        Wed Jul 13 13:59:55 2011 +0200
files:       binr/rax2/rax2.c doc/fortunes libr/fs/Makefile libr/fs/p/grub/Makefile libr/include/r_io.h libr/io/p/io_w32dbg.c
description:
* Fix w32 and w64 build (r_fs/grub issue)
* Dont quit rax2 when reading from stdin until 'q' or ^D
* Minor fix in fortunes


changeset:   1610:2637b35a1b31
user:        pancake
date:        Wed Jul 13 12:21:56 2011 +0200
files:       libr/core/cmd.c
description:
* Fix double pdf (thanks vext01)


changeset:   1609:5b507dbd2c5e
user:        pancake
date:        Wed Jul 13 11:30:25 2011 +0200
files:       Makefile
description:
* Remove manpages in purge
* Fix make symstall


changeset:   1608:980201d7e995
user:        pancake@pair
date:        Wed Jul 13 10:14:30 2011 +0200
files:       binr/binr.mk binr/radare2/Makefile binr/radiff2/Makefile binr/rafind2/Makefile doc/static libr/debug/p/libgdbwrap/gdbwrapper.c libr/debug/p/libgdbwrap/include/gdbwrapper-internals.h libr/debug/p/libgdbwrap/include/gdbwrapper.h libr/debug/t/Makefile libr/hash/t/Makefile libr/io/Makefile libr/io/p/debug.mk libr/io/p/gdb.mk libr/io/p/haret.mk libr/io/p/io_gdb.c libr/io/p/mach.mk libr/io/p/malloc.mk libr/io/p/shm.mk libr/io/t/Makefile libr/socket/t/Makefile mk/gcc.mk
description:
* Fix static build (--without-pic --with-nonpic)


changeset:   1607:3a06619b5226
user:        pancake@pair
date:        Tue Jul 12 23:22:33 2011 +0200
files:       libr/core/asm.c libr/core/cmd.c
description:
* Fix pi and /c commands


changeset:   1606:e220fccf6bc4
user:        earada
date:        Tue Jul 12 23:06:06 2011 +0200
files:       libr/core/cmd.c
description:
* Fix some warnings


changeset:   1605:d15cbb5291a3
user:        pancake
date:        Tue Jul 12 19:42:20 2011 +0200
files:       doc/windows libr/io/p/io_w32dbg.c mk/x86_64-w64-mingw32-gcc.mk
description:
* Fix build for ming-w64


changeset:   1604:026708c1cb88
user:        earada
date:        Tue Jul 12 17:26:56 2011 +0200
files:       libr/core/vmenus.c libr/print/format.c
description:
* Enhacements in keybindings used in visual mounts menus
* Implement missing 'pf p' pointer format


changeset:   1603:71f82c98175a
user:        pancake@pairch

libr/main/rvc2.c
/* radare - LGPL - Copyright 2021 - pancake */
#include <rvc.h>

static void rvc2_show_help(void) {
	printf ("Usage: rvc2 [action] [file ...]\n"
		" init            initialize repository in current directory\n"
		" add [file ..]   add files to the current repository\n"
		" checkout [name] checkout given branch name\n"
		" log             list commits in current branch\n"
		" branch          list all available branches\n"
		" branch [name]   change to another branch\n"
		"Examples:\n"
		"  rvc2 init\n"
		"  man rvc2\n");
}

R_API int r_main_rvc2(int argc, const char **argv) {
	RGetopt opt;
	int c;
	bool git = false;

	r_getopt_init (&opt, argc, argv, "gvh");
	while ((c = r_getopt_next (&opt)) != -1) {
		switch (c) {
		case 'g':
			git = true;
			break;
		case 'v':
			return r_main_version_print ("rvc2");
		case 'h':
			rvc2_show_help ();
			return 0;
		default:
			rvc2_show_help ();
			return -1;
		}
	}

	if (git) {
		eprintf ("TODO: r_vc_git APIs should be called from r_vc\n");
		eprintf ("TODO: r_vc_new should accept options argument\n");
	}
	const char *action = (opt.ind < argc)? opt.arg: NULL;
	if (action) {
		if (!strcmp (action, "init")) {
			char *path = r_sys_getdir ();
			Rvc *vc = r_vc_new (path);
			if (vc) {
				r_vc_free (vc);
				return 0;
			}
			return 1;
		} else if (!strcmp (action, "branch")) {
			char *path = r_sys_getdir ();
			Rvc *vc = r_vc_new (path);
			if (vc) {
				if (opt.ind + 1 < argc) {
					const char *name = argv[opt.ind + 1];
					r_vc_branch (vc, name);
				} else {
					RListIter *iter;
					RvcBranch *b;
					r_list_foreach (vc->branches, iter, b) {
						printf ("%s  %s\n", b->head->hash, b->name);
					}
				}
				r_vc_free (vc);
				return 0;
			}
			return 1;
		} else if (!strcmp (action, "add")) {
			char *path = r_sys_getdir ();
			Rvc *vc = r_vc_new (path);
			if (vc) {
				int i;
				RList *files = r_list_newf (free);
				for (i = opt.ind ; i < argc; i++) {
					r_list_append (files, strdup (argv[i]));
				}
				RList *blobs = r_vc_add (vc, files); // print the blobs?
				RListIter *iter;
				RvcBlob *b;
				r_list_foreach (blobs, iter, b) {
					printf ("%s  %s\n", b->hash, b->fname);
				}
				r_list_free (files);
				r_vc_free (vc);
				return 0;
			}
			return 1;
		}
	}
	return 0;
}
